tmt/steps/provision/__init__.py

```diff
@@ -146,3 +150,231 @@ def guest(self):
+                distro = re.search('DESCRIPTION="(.*)"', distro).group(1)
+            except (tmt.utils.RunError, AttributeError):
+                distro = None
+        if distro:
+            self.info('distro', distro, 'green')
+        # Kernel
+        kernel = self.execute('uname -r', dry=True)[0].strip()
+        self.verbose('kernel', kernel, 'green')
+
+    def _ansible_verbosity(self):
+        """ Prepare verbose level based on the --debug option count """
+        if self.opt('debug') < 3:
+            return ''
+        else:
+            return ' -' + (self.opt('debug') - 2) * 'v'
+
+    def _ansible_summary(self, output):
+        """ Check the output for ansible result summary numbers """
+        if not output:
+            return
+        keys = 'ok changed unreachable failed skipped rescued ignored'.split()
+        for key in keys:
+            matched = re.search(rf'^.*\s:\s.*{key}=(\d+).*$', output, re.M)
+            if matched and int(matched.group(1)) > 0:
+                tasks = fmf.utils.listed(matched.group(1), 'task')
+                self.verbose(key, tasks, 'green')
+
+    def _ansible_playbook_path(self, playbook):
+        """ Prepare full ansible playbook path """
+        # Playbook paths should be relative to the metadata tree root
+        self.debug(f"Applying playbook '{playbook}' on guest '{self.guest}'.")
+        playbook = os.path.join(self.parent.plan.run.tree.root, playbook)
+        self.debug(f"Playbook full path: '{playbook}'", level=2)
+        return playbook
+
+    def _export_environment(self, execute_environment=None):
+        """ Prepare shell export of environment variables """
+        # Prepare environment variables so they can be correctly passed
+        # to ssh's shell. Create a copy to prevent modifying source.
+        environment = dict()
+        environment.update(execute_environment or dict())
+        # Plan environment and variables provided on the command line
+        # override environment provided to execute().
+        environment.update(self.parent.plan.environment)
+        # Prepend with export and run as a separate command.
+        if not environment:
+            return ''
+        return 'export {}; '.format(
+            ' '.join(tmt.utils.shell_variables(environment)))
+
+    def ansible(self, playbook):
+        """ Prepare guest using ansible playbook """
+        playbook = self._ansible_playbook_path(playbook)
+        stdout, stderr = self.run(
+            f'stty cols {tmt.utils.OUTPUT_WIDTH}; ansible-playbook '
+            f'--ssh-common-args="{self._ssh_options(join=True)}" '
+            f'-e ansible_python_interpreter=auto'
+            f'{self._ansible_verbosity()} -i {self._ssh_guest()}, {playbook}')
+        self._ansible_summary(stdout)
+
+    def execute(self, command, **kwargs):
+        """
+        Execute command on the guest
+
+        command ....... string or list of command arguments
```

Comment:
So, the caller is actually responsible for quoting? The command is simply space-joined, so in case my input `list` has entries with spaces this doesn't work.

I'd expect this to be passed to `run()` without the join, or handled in some other way. Or accept only string, and let caller handle that join/quotation explicitly.

Reply:
@psss could you briefly touch on this please?

Reply:
Yes, quote escaping, if needed, has to be handled by the caller. Added a clarification note. The command has to be joined with `ssh` arguments and other stuff such as `cwd` or environment so I'm not sure it's possible to provide a way to handle list of arguments to run. But that's a different question / issue / pull request.